Activity filter counts

carlobeltrame opened the pull request:
Adds counters indicating the number of results belonging to each filter option.
The counters adapt when more filters are added.
There is some complexity due to the different filter behaviours (AND vs OR, exclusive none, single vs. multiple), but I think I got the most intuitive version down for each filter.
Works both on the dashboard and on the picasso.

Also fixes #4951, and I moved the filtering logic to common, because of the following feature request: "Druckfilter (z.B. nur Blöcke wo ich Verantwortlich bin)". This PR is incidentally a first step in this direction.

![image](https://github.com/user-attachments/assets/7d0991cc-a05a-471a-addd-e24b0ddd7c20)

carlobeltrame committed:
Refactor: Use the same filtering logic for dashboard and picasso

carlobeltrame committed:
Fix: Typo in filter initialization

carlobeltrame committed:
Feat: Add counters to filter dropdowns

carlobeltrame committed:
Refactor: Move filter logic to common

This is in anticipation of being able to use the same filtering logic in
the prints in the future, e.g. to filter the printed schedule entries in
a camp programme.

carlobeltrame committed:
Chore: Fix eslint

pmattmann commented on common/helpers/filterMatchScheduleEntry.js:
```diff
@@ -0,0 +1,44 @@
+    (filter.category === null ||
+      filter.category === undefined ||
+      filter.category.length === 0 ||
+      filter.category?.includes(
```
```suggestion
      filter.category.includes(
```

carlobeltrame replied:
[4014a33](https://github.com/ecamp/ecamp3/pull/7376/commits/4014a3356bdf9dbbfa2b640a2b3e64b8f19011cd)

pmattmann commented on common/helpers/filterMatchScheduleEntry.js:
```diff
@@ -0,0 +1,44 @@
+    (filter.responsible === null ||
+      filter.responsible === undefined ||
+      filter.responsible.length === 0 ||
+      typeof filter.responsible?.every !== 'function' ||
+      filter.responsible?.every((responsible) => {
```
```suggestion
      filter.responsible.every((responsible) => {
```

I would remove the `typeof` check. 
a) Function exists - then the check is not needed. 
b) Function does not exist - then the feature does not work.

As it is now, we can't determine if it would need a polyfill.

carlobeltrame replied:
Okay, fixed in [4014a33](https://github.com/ecamp/ecamp3/pull/7376/commits/4014a3356bdf9dbbfa2b640a2b3e64b8f19011cd). I had the function check there for the case that a non-array is passed in the responsible filter, which we currently don't support. This will throw an error without the function check. But it should surface quickly, should it ever happen.

pmattmann commented on common/helpers/filterMatchScheduleEntry.js:
```diff
@@ -0,0 +1,44 @@
+    (filter.progressLabel === null ||
+      filter.progressLabel === undefined ||
+      filter.progressLabel.length === 0 ||
+      filter.progressLabel?.includes(
```
```suggestion
      filter.progressLabel.includes(
```

carlobeltrame replied:
[4014a33](https://github.com/ecamp/ecamp3/pull/7376/commits/4014a3356bdf9dbbfa2b640a2b3e64b8f19011cd)

pmattmann reviewed (approved):
🤩
only minor details

carlobeltrame committed:
Simplify filter function

We ignore the error case when the filter object contains something other
than an array in the responsible key for now. Since it throws an error,
this should quickly surface during development if anyone uses it
wrongly, or at least we will see it in sentry.

manuelmeister commented on frontend/src/views/camp/Dashboard.vue:
```diff
@@ -24,6 +24,10 @@
         :loading-endpoints="loadingEndpoints"
         :camp="camp"
         :periods="periods"
+        :filter-fn="
+          (filter) =>
+            scheduleEntries.filter((scheduleEntry) => filterFn(scheduleEntry, filter))
+        "
```
I'm not sure if this should be a computed function / or if this is automatically made computed

carlobeltrame replied:
[2821b8f](https://github.com/ecamp/ecamp3/pull/7376/commits/2821b8ff4b258a60d982cf76ce368738e248877a)

manuelmeister commented on common/helpers/__tests__/filterMatchScheduleEntry.spec.js:
Could you also please test multiple properties at the same time?

carlobeltrame replied:
[a8c7746](https://github.com/ecamp/ecamp3/pull/7376/commits/a8c77468d0c0e7180eeba2416ed756a4078290cb)

manuelmeister reviewed (approved):
Nice! Good idea
I'm not sure about the visual design.

Maybe this could be "decluttered" with some minor visual clues:
![Bildschirmfoto 2025-05-07 um 22 14 21](https://github.com/user-attachments/assets/709f3dc2-16e7-4ce0-b840-f7d62c6db24b)
But I can test this in a separate PR.

carlobeltrame committed:
Add tests with multiple filter criteria at once

carlobeltrame committed:
Use computed instead of view logic

carlobeltrame committed:
Styling of result count as badge

carlobeltrame commented:
@manuelmeister 
> Maybe this could be "decluttered" with some minor visual clues

Your version looks so much better! I tried to replicate it. I didn't manage to get a satisfying border, I had to use `!important` because vuetify 2 does not allow to set the text color (as opposed to vuetify 3), and I duplicated the same styles in the two filter components... Not sure how you usually handle things like this, would this already go into a global css file now?

Anyways, have a look and feel free to improve the CSS or merge if ready.

manuelmeister committed:
Replace count badge

manuelmeister committed:
Ignore false positive eslint rules

manuelmeister commented:
@carlobeltrame can you have a look at it again? I tried to solve it using a component.

carlobeltrame committed:
Fix filter result counter on mobile picasso
